### PR TITLE
VCU Simulation Framework

### LINKED ISSUE #61: Create VCU Simulation Framework for Integration Testing
**Issue Summary:**
Develop a comprehensive VCU Simulation Framework that empowers users to conduct detailed testing of the vehicle's firmware under race-like scenarios. This framework will facilitate thorough verification of the VCU's behavior in response to various critical inputs and stimuli, including 
-APPS (Accelerator Pedal Position Sensor)
-BSE (Brake System Enable) pedals
-RTD (Ready-To-Drive) signals
-Set-Reset-Flip Events
-TSAL (Transmission Shift Actuator Lock).

**Scope and Purpose:**
The primary goal of this framework is to enable rigorous integration testing of the VCU's behavior. It will provide a controlled environment where the VCU's responses to various inputs can be tested comprehensively. The testing scenarios will simulate real-world conditions, ensuring that the VCU functions correctly under different operational states.

**Key Features:**
- **Realistic Race-Like Scenarios:** The framework will allow users to mimic race-like conditions, creating a dynamic testing environment that closely resembles actual vehicle operation.

- **Multi-Input Simulation:** Users can simulate various critical inputs, such as accelerator and brake pedal actions, RTD signals, and shift actuator behaviors, to assess how the VCU responds.

- **Detailed Event and State Testing:** The framework will offer in-depth testing of VCU events and states, ensuring that the VCU behaves as expected in every operational phase.

**Benefits:**
- **Thorough Verification:** Integration testing with this framework will enable comprehensive verification of the VCU's functionality and ensure that it responds correctly to a wide range of inputs and stimuli.

- **Early Issue Detection:** The ability to simulate real-world scenarios will allow for early detection of potential issues and improve the reliability of the VCU.

- **Robust Validation:** The framework will validate the correctness of the VCU's events and states, enhancing the overall safety and performance of the vehicle.

**Additional Information:**
This issue seeks to create a flexible and effective VCU Simulation Framework that enhances integration testing. Your contributions and ideas are highly valued to ensure the successful implementation of this framework. Please provide your insights and suggestions on how to best achieve this objective.

### PR DESCRIPTION
**Pull request by kevinl03**
# Overview

Develop a comprehensive VCU Simulation Framework that empowers users to conduct detailed testing of the vehicle's firmware under race-like scenarios. This framework will facilitate thorough verification of the VCU's behavior in response to various critical inputs and stimuli, including
>APPS (Accelerator Pedal Position Sensor)
>BSE (Brake System Enable) pedals
>RTD (Ready-To-Drive) signals
>Set-Reset-Flip Events
>TSAL (Transmission Shift Actuator Lock)

## Development discussion

A detailed discussion about the features and scope of the framework can be found via this github issue https://github.com/sfuphantom/vcu-fw/issues/61

## User Manual

Refer to the following document for environment set up and framework usage
https://docs.google.com/document/d/1h_xk5TsREIRYUv_hB6RO2n2hTzEec8QP7Y_7VT4gCtU/edit 

## Design Document

The following document outlines the framework design decisions for the VCU testing robustness. Developers who wish to modify the current usage should refer to 

https://docs.google.com/document/d/10IChPywZE57cQT93kg4YBlf_Gsr4lBgPxA6Phs91sME/edit

## High level architecture 

image

## Framework implementation:

VCU Test Framework UML

### CODE REVIEW

File: command-line/data_generation.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+class DataGeneration(CsvGeneration, PlotGeneration):
+    """
+    This class inherits functionality from both CsvGeneration and PlotGeneration and provides combined
+    capabilities for generating plots and writing data to CSV files.
+    """
+    def __init__(self) -> None:
+        super().__init__()
+
+
+# import matplotlib.pyplot as plt
```
Comment (kkramer215) [inline command-line/data_generation.py]: If this section is not needed in the future, probably just get rid of it or add a TODO comment if it's a WIP.

File: command-line/sim_interface.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        for sim_num in range(sim_len):
+
+            apps1: float = sim_model[VCU_Pedals.APPS1][sim_num]
+            apps2: float = sim_model[VCU_Pedals.APPS2][sim_num]
+            bse: float = sim_model[VCU_Pedals.BSE][sim_num]
+
+            end_time = time.time()
+
+            
+            # response_vcu: ResponseVCU = ResponseVCU(self.vcu_writer.write(int(apps1), int(apps2), int(bse)))
```
Comment (kkramer215) [inline command-line/sim_interface.py]: Are these lines for the user to uncomment or extra code that needs cleanup?

**Review by kkramer215 — CHANGES_REQUESTED** [review]
Comments for the functions and classes are very helpful while following along with the code. Just make a note for each of the comments in the files and resolve any merge conflicts but looks really good!

**Review by rafguevara14 — CHANGES_REQUESTED** [review]
Consider that with regard to readability and maintainability, the intention of your code should be front and center, and be as obvious as possible. In some places, the use of type annotations dominates the code more than they would in a strongly typed language like C++.

**Long type annotations:**

These types with nested dicts and lists should either be declared as its own class or type aliased. The information is still relevant, but we should only have to read the super-long definition when we need to.
```
# you have to get past the type annotation to see the actual intention of the code
write_res : dict[Union[VCU_Pedal, str], Union[list[int], list[EventData], list[StateData]]] = self._write_data()
-------------------
# at top of file
VCUResult = dict[Union[VCU_Pedal, str], Union[list[int], list[EventData], list[StateData]]] 

write_res : VCUResult = self._write_data()
```
 
**Redundant annotations:**
Adding type annotations for every single line of code can add unnecessary noise to the codebase; not all situations need strict typing. This is a little more subjective about where you choose to forego type annotation.

A good rule to follow is to type annotate function arguments and return values. Types can then be deduced from these return values. If it's not obvious, then add a type annotation.

```
json_config: dict = self._parse_json()
def _parse_json(self) -> Union[dict, None]:

-------
# the IDE and the programmer can deduce the type easily from this simple assignment. 
json_config = self._parse_json() 
```


**Incorrect, misleading, and/or non-obvious:**
```
interface: VCUSimInterface = VCUSimInterface().begin() # the begin method does not actually return a VCUSimInterface
```

Also found a [stack overflow post ](https://stackoverflow.com/questions/51210914/type-hints-when-to-annotate)that I 100% agree with giving more advice on type hints
image

File: command-line/sim_interface.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+         Attributes:
+            vcu_writer (VCUSimulation): An instance of the VCUSimulation class for VCU serial communication.
+
+            simulation (Simulation): An instance of the Simulation class for simulation generation
+
+            vcu_gdrive_interface (VcuGDriveInterface): An instance of the VcuGDriveInterface class for interfacing with G Drive Folders
+
+            data_generator (DataGeneration): An reference to the DataGeneration class for data formatting.
+        """
+        self.vcu_writer = VCUSimulation
```
Comment (rafguevara14) [inline command-line/sim_interface.py]: A little confused about how this is used. You assign self.vcu_writer as the **literal type VCUSimulation** (not an instance of VCUSimulation) here, but refer to it elsewhere as an instance of VCUSimulation.

File: command-line/sim_interface.py (hunk 2 above)
Comment (kevinl03) [reply command-line/sim_interface.py]: Fixed, they art part of the framework execution and should have been uncommented. I accidentally commited these changes while testing the formatting

bfcd917f06f4cbad3e850ea23e66865096742888

File: command-line/sim_interface.py (hunk 3 above)
Comment (kevinl03) [reply command-line/sim_interface.py]: A reference is passed because we want to have functionality to configure the device if we choose to add more attributes in the constructor. 

These attributes would live in the configs/device_config.json file. When the device is parsed, an instance with the config creates an instance of the writer.

I don't personally like this structure very much either way. I think the config should be parsed within the VCU Class, not the interface. I chose not to touch the VCUCommunication class (previously you had it as VCUSimulation) since it was your initial design. 

I.e I would rather have `self.configure_device` live within the VCUSimulation class, and parsed during the constructor. 

>>> VCUSimulation(config : dict = None) 
>>>      if config = None:
>>>          self.parsejson()
>>>      else:
>>>           self.Find_Port()
>>>           self.baudrate = self.BAUDRATE

Comment (rafguevara14) [reply command-line/sim_interface.py]: Ah I see, you're temporarily assigning it the class type and then initializing it later once you have the correct information...**And then** passing it to the Simulation class where it seems like it will never be used by `VCUSimInterface` again. That's a lot of redirection for a simple assignment; up until we parse the JSON, the type annotations are wrong and the ownership/scope of vcu_writer is unclear.

I see why you want  `VCUSimInterface` to have vcu_writer, but having multiple owners of a shared resource is generally a bad idea. Synchronization logic would be spread out across the owners making it harder to manage and maintain. A better design would be a class that protects the shared resource and orchestrates the synchronization. Sort of like the Proxy design pattern. If we decide to test manual control while running a sim model, we'd put the vcu_writer under this proxy.

For now, `VCUSimInterface` is the final class that is composed of all of the components/objects in the test framework. Any internal details of those classes (like vcu_writer) should not be its responsibility. It also seems like a lot of one liner functions. In the future these might get bigger, but we shouldn't prematurely optimize before that.

```
 def __init__(self) -> None:
        """
        Initialize all components for simulation communication including
        Drive Communication, VCU serial writing, Simulation Generation, and Data Formatting
        """
        
        json_config = self._parse_json()
        vcu_writer = device(**json_config)
        
        #Pass in VCUSimulaiton object as reference for thread-lock compatibility.
        #Useful if we decide to test manual control while running a sim model
        self.simulation : Simulation = Simulation(vcu_writer)
        self.vcu_gdrive_interface : VcuGDriveInterface = VcuGDriveInterface()
        self.data_generator = DataGeneration
```

Also, what you refer to as references throughout the code is just passing the class. So data_generator is a class not a reference. I don't think references are a solid concept in Python that you can really control so commenting on it can be confusing.

Comment (rafguevara14) [reply command-line/sim_interface.py]: I agree, the structure is a little weird...not sure about putting it in the VCUSimulation class but we can make an issue for any smaller design choices we want to deal with later

Comment (kevinl03) [reply command-line/sim_interface.py]: In the current design, both the `VCUSimInterface` and the `Simulation` class share an instance of the VCU writer. This instance is created using the `set_metadata` function within the SimInterface.

The `Simulation` class, where the majority of the command-line interface is implemented, presents two options: Manual Control and Simulation Generation.

For Manual Control, the simulation writer is directly utilized. Users can input a 6-command argument: `<APPS1> <APPS2> <BSE> <TSAL> <RTD> <SETRESET>`, which is then transmitted to the VCU.

Simulation Generation takes in different arguments: `<Cycles> <Precision> <APPS Wave> <BSE wave>`. Once a simulation is created (potentially comprising multiple simulation models appended together), the user calls execute. At this point, the responsibility shifts to the `VCUSimInterface` to send the complete simulation to the VCU, referencing the `_write_data(self)` method.

Regarding the suggestion to introduce a thread lock for concurrent Manual Control and Simulation execution: This is to ensure that while a simulation is being executed, the command-line interface remains responsive to user inputs. This is crucial due to the VCU's soft-limit of 33ms, where any delay beyond this might jeopardize data transmission reliability.

Upon reflection, it's clear that only the `VCUSimInterface` should manage the `VCU_Communication` instance (referred to as the `vcu writer`). We propose two changes:

1. **Move configuration responsibilities:** The configuration of `parse_json` and `set_metadata` should be within the `VCU_Communication` class. This shift alleviates the `VCUSimInterface` from device configuration duties.

2. **Define command-line inputs within VCUSimInterface:** The command line inputs should be defined within the `VCUSimInterface`. This allows us to eliminate the need to pass the `VCU_Communication` (vcu_writer) object to the `Simulation` class instance in the constructor. Users can then call Manual Control `<APPS1> <APPS2> <BSE> <etc.>`, and the `VCUSimInterface` will invoke the already defined `vcu_writer` instance.

In conclusion, the `VCUSimInterface` will be the sole manager of the `vcu writer`, handling both Manual Control and writing the sim model to the VCU. Prior to this change, the `Simulation` class was responsible for Manual Control due to the command line being defined within it. Additionally, there will be no device configuration clutter within the interface; instead, device configurations will be executed through the constructor of the `VCU_Communication` (during vcu writer initialization).

Comment (kevinl03) [reply command-line/sim_interface.py]: This way, the `Proxy` would also reside within the VCUSimInterface, or we just keep it as a seperate entitiy. Overall, this means managing the writing device would be a lot more simple.

File: command-line/data_generation.py (hunk 1 above)
Comment (kevinl03) [reply command-line/data_generation.py]: removed for now, it is a WIP in a different file that is not commited. If I get a change, I will add into the DataGeneration Class. It was a way to display a live graph of the current values being send to the VCU

b69e81cadbde6a3e8f587e39e2b2141e8ed4ae0b

File: command-line/sim_interface.py (hunk 3 above)
Comment (kevinl03) [reply command-line/sim_interface.py]: ## TL;DR

1. **VCU Communication Class Enhancement:**
   - The `VCU Communication` class now takes responsibility for configuring the VCU writer device from JSON if no port is provided in the constructor.
   - [Commit Link](https://github.com/sfuphantom/vcu-fw/commit/50dfb8fbb82cd14a4aee6175028071567087af84)

2. **Interface Class Refinement:**
   - Command-line inputs are now defined within the Interface class, removing the need for the simulation generator class to contain a VCU writer instance (`VCU Communication` class).
   - [Commit Link](https://github.com/sfuphantom/vcu-fw/commit/0333c7de23e14cdf513d839352bbdbd89d54da86)

## Clean Separation of Responsibilities in Interface Class

Clean Separation

Comment (kevinl03) [reply command-line/sim_interface.py]: Lastly, I fully agree with you with very long type annotations. That is now fixed

e1c98fe5ebcdd279fccbe6248d0cbb927646c5f8

Comment (rafguevara14) [reply command-line/sim_interface.py]: Yup! Agree with the vcu_writer changes. resolving thread

**Review by rafguevara14 — APPROVED** [review]
This should be a good baseline for the project. Now we can move this over to its own repo and continue development there

**Review by kkramer215 — APPROVED** [review]
